ada_feeding/ada_feeding/behaviors/moveit2/moveit2_plan.py

```diff
@@ -0,0 +1,641 @@
+        end_joint_state (JointState): get the last joint state of the planned trajectoy,
+                                      useful for chaining planning calls
+        """
+        # pylint: disable=unused-argument, duplicate-code
+        # Arguments are handled generically in base class.
+        super().blackboard_outputs(
+            **{key: value for key, value in locals().items() if key != "self"}
+        )
+
+    @override
+    def setup(self, **kwargs):
+        # Docstring copied from @override
+
+        # pylint: disable=attribute-defined-outside-init
+        # It is okay for attributes in behaviors to be
+        # defined in the setup / initialise functions.
+
+        # Get Node from Kwargs
+        self.node = kwargs["node"]
+
+        # Get the MoveIt2 object.
+        self.moveit2, self.moveit2_lock = get_moveit2_object(
+            self.blackboard,
+            self.node,
+        )
+
+        # Get TF Listener from blackboard
+        self.tf_buffer, _, self.tf_lock = get_tf_object(self.blackboard, self.node)
+
+    @override
+    def initialise(self) -> None:
+        # Docstring copied from @override
+
+        # pylint: disable=attribute-defined-outside-init
+        # It is okay for attributes in behaviors to be
+        # defined in the setup / initialise functions.
+
+        self.planning_future = None
+
+        # Copy config to MoveIt2 object.
+        # Note: this *could* block, but it is unlikely
+        with self.moveit2_lock:
+            self.moveit2.planner_id = self.blackboard_get("planner_id")
+            self.moveit2.pipeline_id = self.blackboard_get("pipeline_id")
+            self.moveit2.max_velocity = self.blackboard_get("max_velocity_scale")
+            self.moveit2.max_acceleration = self.blackboard_get(
+                "max_acceleration_scale"
+            )
+            self.moveit2.allowed_planning_time = self.blackboard_get(
+                "allowed_planning_time"
+            )
+            self.moveit2.cartesian_jump_threshold = self.blackboard_get(
+                "cartesian_jump_threshold"
+            )
+            # If the plan is cartesian, it should always avoid collisions
+            self.moveit2.cartesian_avoid_collisions = True
+
+            # Clear constraints
+            self.moveit2.clear_goal_constraints()
+            self.moveit2.clear_path_constraints()
+
+    @override
+    def update(self) -> py_trees.common.Status:
+        # Docstring copied from @override
+
+        # pylint: disable=too-many-return-statements, too-many-statements, too-many-branches
+        # All raised exceptions map to a return statement.
+        # This is the main function, it is okay for it to be long.
+
+        # Lock MoveIt2 Object
+        if self.moveit2_lock.locked():
+            return py_trees.common.Status.RUNNING
+        if self.tf_lock.locked():
+            return py_trees.common.Status.RUNNING
+        with self.moveit2_lock, self.tf_lock:
```

Comment:
The problem is that with statement need to be preceded by an "if locked return running" to prevent blocking the tick thread.

But that could cause update to return prematurely and lose or recalculate local state.

I think it's better to treat the entire update, which should be quick anyway, with a single with block to prevent this.

Reply:
Good point, I agree with the way you’ve currently written the code.